amr-wind/fvm/qcriterion_nondim.H

```diff
@@ -0,0 +1,126 @@
+        : m_qcritnomdimphi(qcritnomdimphi), m_phi(phi)
+    {
+        AMREX_ALWAYS_ASSERT(AMREX_SPACEDIM == m_phi.num_comp());
+    }
+
+    template <typename Stencil>
+    void apply(const int lev, const amrex::MFIter& mfi) const
+    {
+        const auto& geom = m_phi.repo().mesh().Geom(lev);
+        const auto& idx = geom.InvCellSizeArray();
+        const auto& qcritnomdimphi = m_qcritnomdimphi(lev).array(mfi);
+        const auto& phi = m_phi(lev).const_array(mfi);
+
+        const auto& bx_in = mfi.tilebox();
+        const auto& bx = Stencil::box(bx_in, geom);
+        if (bx.isEmpty()) return;
+        amrex::ParallelFor(
+            bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                amrex::Real cp1, c, cm1, ux, uy, uz, vx, vy, vz, wx, wy, wz;
+                cp1 = Stencil::c00;
+                c = Stencil::c01;
+                cm1 = Stencil::c02;
+
+                ux = (cp1 * phi(i + 1, j, k, 0) + c * phi(i, j, k, 0) +
+                      cm1 * phi(i - 1, j, k, 0)) *
+                     idx[0];
+                vx = (cp1 * phi(i + 1, j, k, 1) + c * phi(i, j, k, 1) +
+                      cm1 * phi(i - 1, j, k, 1)) *
+                     idx[0];
+                wx = (cp1 * phi(i + 1, j, k, 2) + c * phi(i, j, k, 2) +
+                      cm1 * phi(i - 1, j, k, 2)) *
+                     idx[0];
+
+                cp1 = Stencil::c10;
+                c = Stencil::c11;
+                cm1 = Stencil::c12;
+
+                uy = (cp1 * phi(i, j + 1, k, 0) + c * phi(i, j, k, 0) +
+                      cm1 * phi(i, j - 1, k, 0)) *
+                     idx[1];
+                vy = (cp1 * phi(i, j + 1, k, 1) + c * phi(i, j, k, 1) +
+                      cm1 * phi(i, j - 1, k, 1)) *
+                     idx[1];
+                wy = (cp1 * phi(i, j + 1, k, 2) + c * phi(i, j, k, 2) +
+                      cm1 * phi(i, j - 1, k, 2)) *
+                     idx[1];
+
+                cp1 = Stencil::c20;
+                c = Stencil::c21;
+                cm1 = Stencil::c22;
+
+                uz = (cp1 * phi(i, j, k + 1, 0) + c * phi(i, j, k, 0) +
+                      cm1 * phi(i, j, k - 1, 0)) *
+                     idx[2];
+                vz = (cp1 * phi(i, j, k + 1, 1) + c * phi(i, j, k, 1) +
+                      cm1 * phi(i, j, k - 1, 1)) *
+                     idx[2];
+                wz = (cp1 * phi(i, j, k + 1, 2) + c * phi(i, j, k, 2) +
+                      cm1 * phi(i, j, k - 1, 2)) *
+                     idx[2];
+
+                const amrex::Real S2 =
+                    2.0 * std::pow(ux, 2) + 2.0 * std::pow(vy, 2) +
+                    2.0 * std::pow(wz, 2) + std::pow(uy + vx, 2) +
+                    std::pow(vz + wy, 2) + std::pow(wx + uz, 2);
+
+                const amrex::Real W2 = std::pow(uy - vx, 2) +
+                                       std::pow(vz - wy, 2) +
+                                       std::pow(wx - uz, 2);
```

Comment:
This code looks like a duplicate of the QCriterion class. Can these not be combined together and differentiated at the operator functions that are defined below?